aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -2763,10 +2764,20 @@ Tensor linalg_svdvals(const Tensor& input) {
       "torch.linalg.svdvals: input should have at least 2 dimensions, but has ",
       input.dim(),
       " dimensions instead");
+
   Tensor singular_values;
-  std::tie(std::ignore, singular_values, std::ignore) =
-      // NOLINTNEXTLINE(bugprone-argument-comment)
-      at::_svd_helper(input, /*full_matrices=*/false, /*compute_uv=*/false);
+
+  // if input requires grad we must compute the singular vectors to make this function differentiable
+  // the singular vectors are not exposed to the user
+  if (torch::autograd::compute_requires_grad(input)) {
+    std::tie(std::ignore, singular_values, std::ignore) =
+        // NOLINTNEXTLINE(bugprone-argument-comment)
+        at::_svd_helper(input, /*!full_matrices=*/true, /*compute_uv=*/true);
+  } else {
+    std::tie(std::ignore, singular_values, std::ignore) =
+        // NOLINTNEXTLINE(bugprone-argument-comment)
+        at::_svd_helper(input, /*!full_matrices=*/false, /*compute_uv=*/false);
+  }
```

Comment:
Perhaps is simpler to avoid repeating code:
```c++
const auto needs_grad = torch::autograd::compute_requires_grad(input);
return std::get<1>(at::_svd_helper(input, /*!full_matrices=*/needs_grad, /*compute_uv=*/needs_grad));
``` 
Also, can't we just pass `!full_matrices = True` in both cases?

Reply:
`!full_matrices` doesn't do anything if `compute_uv=False`, so it doesn't matter, it can be true in both code paths.

If the singular vectors are computed in full matrix form, then there is some additional work in the derivative function
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/FunctionsManual.cpp#L2066-L2078

Reply:
That's what I thought. Then, I update my proposal to:
```c++
const auto compute_uv = torch::autograd::compute_requires_grad(input);
return std::get<1>(at::_svd_helper(input, /*!full_matrices=*/True, /*compute_uv=*/compute_uv));
```

Reply:
I find std::tie with std::ignore more clear than std::get<1>. I can of course change that, it doesn't matter now.

Reply:
Whichever works best. I'd still remove the if/else logic though